lib/ui/channel_buffers.dart

```diff
@@ -0,0 +1,173 @@
+  ///
+  /// Size 1 implies that you only care about the most recent value.
+  ///
+  /// Size >1 means you want to process every single message and want to chose a
+  /// buffer size that will avoid any overflows.
+  static const int kDefaultBufferSize = 1;
+
+  final Map<String, _RingBuffer<_StoredMessage>> _messages =
+    <String, _RingBuffer<_StoredMessage>>{};
+
+  _RingBuffer<_StoredMessage> _makeRingBuffer(int size) {
+    final _RingBuffer<_StoredMessage> result = _RingBuffer<_StoredMessage>(size);
+    result.dropItemCallback = _onDropItem;
+    return result;
+  }
+
+  void _onDropItem(_StoredMessage message) {
+    message.callback(null);
+  }
+
+  /// Returns true on overflow.
+  bool push(String channel, ByteData data, PlatformMessageResponseCallback callback) {
+    _RingBuffer<_StoredMessage> queue = _messages[channel];
+    if (queue == null) {
+      queue = _makeRingBuffer(kDefaultBufferSize);
+      _messages[channel] = queue;
+    }
+    final bool result = queue.push(_StoredMessage(data, callback));
+    if (result) {
+      // TODO(aaclarke): Update this message to include instructions on how to resize
+      // the buffer once that is available to users.
+      _Logger._printString('Overflow on channel: $channel.  '
+                           'Messages on this channel are being sent faster '
+                           'than they are being processed which is resulting '
+                           'in the dropping of messages.  The engine may not be '
+                           'running or you need to adjust the buffer size.');
+    }
+    return result;
+  }
+
+  /// Returns null on underflow.
+  _StoredMessage _pop(String channel) {
+    final _RingBuffer<_StoredMessage> queue = _messages[channel];
+    final _StoredMessage result = queue?.pop();
+    return result;
+  }
+
+  bool _isEmpty(String channel) {
+    final _RingBuffer<_StoredMessage> queue = _messages[channel];
+    return (queue == null) ? true : queue.isEmpty;
+  }
+
+  void resize(String channel, int newSize) {
+    _RingBuffer<_StoredMessage> queue = _messages[channel];
+    if (queue == null) {
+      queue = _makeRingBuffer(newSize);
+      _messages[channel] = queue;
+    } else {
+      final int numberOfDroppedMessages = queue.resize(newSize);
+      if (numberOfDroppedMessages > 0) {
+        _Logger._printString('Dropping messages on channel "$channel" as a result of shrinking the buffer size.');
+      }
+    }
+  }
+
+  /// Remove and process all stored messages for a given channel.
+  ///
+  /// This should be called once a channel is prepared to handle messages
+  /// (ie when a message handler is setup in the framework).
+  Future<void> drain(String channel, DrainChannelCallback callback) async {
+    while (!_isEmpty(channel)) {
+      final _StoredMessage message = _pop(channel);
+      await callback(message.data, message.callback);
+    }
+  }
+}
+
+final ChannelBuffers channelBuffers = ChannelBuffers();
```

Comment:
Won't each engine have its own isolate and therefore its own channelBuffers?  If this is true this will also be a problem for `ui.window` which I followed.

Reply:
Yup you're right, that was a nonsensical comment :D